### PR TITLE
Optimize: add EncodeTransferMedataRestTemplateInterceptor to RestTemp…

### PR DESCRIPTION
**Pull request by DerekYRC**
## PR Type

Refactoring

## Describe what this PR does for and how you did.

Optimize: add EncodeTransferMedataRestTemplateInterceptor to RestTemplate.

优化代码：为RestTemplate添加拦截器EncodeTransferMedataRestTemplateInterceptor。

## Adding the issue link (#xxx) if possible.

https://github.com/Tencent/spring-cloud-tencent/issues/431

## Note

参照LoadBalancerAutoConfiguration中为RestTemplate添加LoadBalancerInterceptor拦截器的做法，在配置类中注入所有RestTemplate，然后为其添加EncodeTransferMedataRestTemplateInterceptor拦截器

## Checklist

- [x] Add copyright holder at the beginning of .class file if it is new.
- [x] Add information of this PR to CHANGELOG.md in root of project.
- [x] All junit tests passing.
- [x] Coverage from `Codecov Report` should not decrease.

## Checklist (Optional)

- [x] Will Pull Request to branch of 2020.0 and 2021.0.
- [x] Add documentation in javadoc in code or comment below the PR if necessary.
- [x] Add your name as @author to the beginning of .class file.

### CODE REVIEW

File: spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/config/MetadataTransferAutoConfiguration.java [hunk 1]
```diff
@@ -143,69 +141,23 @@ public EncodeTransferMedataFeignInterceptor encodeTransferMedataFeignInterceptor
...
-					// Avoid setting interceptor repeatedly.
-					if (!interceptors.contains(encodeTransferMedataRestTemplateInterceptor)) {
-						interceptors.add(this.encodeTransferMedataRestTemplateInterceptor);
-						restTemplate.setInterceptors(interceptors);
-					}
-				}
-				return bean;
-			}
+		public SmartInitializingSingleton addEncodeTransferMedataInterceptorForRestTemplate(EncodeTransferMedataRestTemplateInterceptor interceptor) {
+			return () -> restTemplates.forEach(restTemplate -> {
```
Comment (SkyeBeFreeman) [inline spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/config/MetadataTransferAutoConfiguration.java:156]: 是不是应该在`BeanPostProcessor`的阶段在做这个操作呢？

Comment (DerekYRC) [reply spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/config/MetadataTransferAutoConfiguration.java:156]: 不是必须在BeanPostProcessor阶段处理。如图一，实例化顺序自上而下，最后SmartInitializingSingleton实例化时会执行其afterSingletonsInstantiated方法（DefaultListableBeanFactory#preInstantiateSingletons）。

这个添加拦截器的逻辑跟LoadBalancerAutoConfiguration中为RestTemplate添加LoadBalancerInterceptor拦截器的做法保持一致。
image
image

Comment (DerekYRC) [reply spring-cloud-starter-tencent-metadata-transfer/src/main/java/com/tencent/cloud/metadata/config/MetadataTransferAutoConfiguration.java:156]: > 是不是应该在`BeanPostProcessor`的阶段在做这个操作呢？

效果一样，区别是BeanPostProcessor是在bean实例化过程中操作，SmartInitializingSingleton是在所有单例bean实例化后操作。

**Review by SkyeBeFreeman — APPROVED** [review]
LGTM.

**Comment by SkyeBeFreeman** [comment]
Can you cherry-pick to branch of 2020 and 2021?

**Comment by DerekYRC** [comment]
> Can you cherry-pick to branch of 2020 and 2021?

会的哈

**Pull request merged**